File: src/algorithms/graph/connectivity/articulation-points/sources/articulation-points_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
articulation_points_module = importlib.import_module("articulation-points")
find_articulation_points = articulation_points_module.find_articulation_points


def test_finds_two_articulation_points_in_default_7_node_graph():
    adjacency_list = {
        "A": ["B", "C"],
        "B": ["A", "C"],
        "C": ["A", "B", "D"],
        "D": ["C", "E", "F"],
        "E": ["D", "G"],
        "F": ["D", "G"],
        "G": ["E", "F"],
    }
    node_ids = ["A", "B", "C", "D", "E", "F", "G"]
    result = find_articulation_points(adjacency_list, node_ids)
    assert set(result) == {"C", "D"}, f"Expected {{C, D}}, got {set(result)}"


def test_returns_no_articulation_points_for_triangle():
    adjacency_list = {
        "A": ["B", "C"],
        "B": ["A", "C"],
        "C": ["A", "B"],
    }
    result = find_articulation_points(adjacency_list, ["A", "B", "C"])
    assert len(result) == 0, f"Expected empty, got {result}"


def test_finds_single_articulation_point_in_path_graph():
    adjacency_list = {
        "A": ["B"],
        "B": ["A", "C"],
        "C": ["B"],
    }
    result = find_articulation_points(adjacency_list, ["A", "B", "C"])
    assert set(result) == {"B"}, f"Expected {{B}}, got {set(result)}"


def test_finds_multiple_articulation_points_in_longer_path():
    adjacency_list = {
        "A": ["B"],
        "B": ["A", "C"],
        "C": ["B", "D"],
        "D": ["C"],
    }
    result = find_articulation_points(adjacency_list, ["A", "B", "C", "D"])
    assert set(result) == {"B", "C"}, f"Expected {{B, C}}, got {set(result)}"


def test_returns_no_articulation_points_for_single_node():
    result = find_articulation_points({"A": []}, ["A"])
    assert len(result) == 0, f"Expected empty, got {result}"


def test_returns_no_articulation_points_for_fully_connected_graph():
    adjacency_list = {
        "A": ["B", "C", "D"],
        "B": ["A", "C", "D"],
        "C": ["A", "B", "D"],
        "D": ["A", "B", "C"],
    }
    result = find_articulation_points(adjacency_list, ["A", "B", "C", "D"])
    assert len(result) == 0, f"Expected empty, got {result}"


def test_finds_star_center_as_articulation_point():
    adjacency_list = {
        "Center": ["A", "B", "C"],
        "A": ["Center"],
        "B": ["Center"],
        "C": ["Center"],
    }
    result = find_articulation_points(adjacency_list, ["Center", "A", "B", "C"])
    assert set(result) == {"Center"}, f"Expected {{Center}}, got {set(result)}"


def test_handles_disconnected_graphs_with_no_articulation_points():
    adjacency_list = {
        "A": ["B", "C"],
        "B": ["A", "C"],
        "C": ["A", "B"],
        "D": ["E", "F"],
        "E": ["D", "F"],
        "F": ["D", "E"],
    }
    result = find_articulation_points(adjacency_list, ["A", "B", "C", "D", "E", "F"])
    assert len(result) == 0, f"Expected empty, got {result}"


if __name__ == "__main__":
    test_finds_two_articulation_points_in_default_7_node_graph()
    test_returns_no_articulation_points_for_triangle()
    test_finds_single_articulation_point_in_path_graph()
    test_finds_multiple_articulation_points_in_longer_path()
    test_returns_no_articulation_points_for_single_node()
    test_returns_no_articulation_points_for_fully_connected_graph()
    test_finds_star_center_as_articulation_point()
    test_handles_disconnected_graphs_with_no_articulation_points()
    print("All tests passed!")

File: src/algorithms/graph/connectivity/articulation-points/__tests__/ArticulationPoints_test.java
import java.util.*;

// Compile: javac ArticulationPoints.java ArticulationPoints_test.java
// Run:     java -ea ArticulationPoints_test
public class ArticulationPoints_test {
    public static void main(String[] args) {
        testFindsTwoArticulationPointsInDefault7NodeGraph();
        testReturnsNoArticulationPointsForTriangle();
        testFindsSingleArticulationPointInPathGraph();
        testFindsMultipleArticulationPointsInLongerPath();
        testReturnsNoArticulationPointsForSingleNode();
        testReturnsNoArticulationPointsForFullyConnectedGraph();
        testFindsStarCenterAsArticulationPoint();
        testHandlesDisconnectedGraphsWithNoArticulationPoints();
        System.out.println("All tests passed!");
    }

    static Map<String, List<String>> adj(String... keyValues) {
        Map<String, List<String>> map = new LinkedHashMap<>();
        for (int idx = 0; idx < keyValues.length; idx += 2) {
            List<String> neighbors = new ArrayList<>();
            for (String neighbor : keyValues[idx + 1].split(",")) {
                if (!neighbor.isEmpty()) neighbors.add(neighbor.trim());
            }
            map.put(keyValues[idx], neighbors);
        }
        return map;
    }

    static void testFindsTwoArticulationPointsInDefault7NodeGraph() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B", "C"));
        adjacencyList.put("B", Arrays.asList("A", "C"));
        adjacencyList.put("C", Arrays.asList("A", "B", "D"));
        adjacencyList.put("D", Arrays.asList("C", "E", "F"));
        adjacencyList.put("E", Arrays.asList("D", "G"));
        adjacencyList.put("F", Arrays.asList("D", "G"));
        adjacencyList.put("G", Arrays.asList("E", "F"));
        List<String> nodeIds = Arrays.asList("A", "B", "C", "D", "E", "F", "G");

        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, nodeIds);
        Set<String> resultSet = new HashSet<>(result);
        assert resultSet.equals(new HashSet<>(Arrays.asList("C", "D"))) :
            "Expected {C, D}, got " + resultSet;
    }

    static void testReturnsNoArticulationPointsForTriangle() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B", "C"));
        adjacencyList.put("B", Arrays.asList("A", "C"));
        adjacencyList.put("C", Arrays.asList("A", "B"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A", "B", "C"));
        assert result.isEmpty() : "Expected empty, got " + result;
    }

    static void testFindsSingleArticulationPointInPathGraph() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B"));
        adjacencyList.put("B", Arrays.asList("A", "C"));
        adjacencyList.put("C", Arrays.asList("B"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A", "B", "C"));
        assert new HashSet<>(result).equals(new HashSet<>(Arrays.asList("B"))) :
            "Expected {B}, got " + result;
    }

    static void testFindsMultipleArticulationPointsInLongerPath() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B"));
        adjacencyList.put("B", Arrays.asList("A", "C"));
        adjacencyList.put("C", Arrays.asList("B", "D"));
        adjacencyList.put("D", Arrays.asList("C"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A", "B", "C", "D"));
        assert new HashSet<>(result).equals(new HashSet<>(Arrays.asList("B", "C"))) :
            "Expected {B, C}, got " + result;
    }

    static void testReturnsNoArticulationPointsForSingleNode() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Collections.emptyList());
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A"));
        assert result.isEmpty() : "Expected empty, got " + result;
    }

    static void testReturnsNoArticulationPointsForFullyConnectedGraph() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B", "C", "D"));
        adjacencyList.put("B", Arrays.asList("A", "C", "D"));
        adjacencyList.put("C", Arrays.asList("A", "B", "D"));
        adjacencyList.put("D", Arrays.asList("A", "B", "C"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A", "B", "C", "D"));
        assert result.isEmpty() : "Expected empty, got " + result;
    }

    static void testFindsStarCenterAsArticulationPoint() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("Center", Arrays.asList("A", "B", "C"));
        adjacencyList.put("A", Arrays.asList("Center"));
        adjacencyList.put("B", Arrays.asList("Center"));
        adjacencyList.put("C", Arrays.asList("Center"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("Center", "A", "B", "C"));
        assert new HashSet<>(result).equals(new HashSet<>(Arrays.asList("Center"))) :
            "Expected {Center}, got " + result;
    }

    static void testHandlesDisconnectedGraphsWithNoArticulationPoints() {
        Map<String, List<String>> adjacencyList = new LinkedHashMap<>();
        adjacencyList.put("A", Arrays.asList("B", "C"));
        adjacencyList.put("B", Arrays.asList("A", "C"));
        adjacencyList.put("C", Arrays.asList("A", "B"));
        adjacencyList.put("D", Arrays.asList("E", "F"));
        adjacencyList.put("E", Arrays.asList("D", "F"));
        adjacencyList.put("F", Arrays.asList("D", "E"));
        ArticulationPoints ap = new ArticulationPoints();
        List<String> result = ap.findArticulationPoints(adjacencyList, Arrays.asList("A", "B", "C", "D", "E", "F"));
        assert result.isEmpty() : "Expected empty, got " + result;
    }
}
